indexing-service/src/main/java/io/druid/indexing/overlord/http/OverlordResource.java

```diff
@@ -851,11 +1071,57 @@ public TaskStatusPlus apply(TaskRunnerWorkItem workItem)
           )
       );
     };
+    Collection<? extends TaskRunnerWorkItem> dataSourceFilter = collectionToFilter;
+    if (dataSource != null) {
+      dataSourceFilter = collectionToFilter
+          .stream()
+          .filter(task -> task.getDataSource().equals(dataSource))
+          .collect(Collectors.toList());
+    }
+    return Lists.newArrayList(
+        AuthorizationUtils.filterAuthorizedResources(
+            req,
+            dataSourceFilter,
+            raGenerator,
+            authorizerMapper
+        )
+    );
+  }
+
+  private List<TaskStatusPlus> securedTaskStatusPlus(
+      List<TaskStatusPlus> collectionToFilter, String dataSource,
+      HttpServletRequest req
+  )
+  {
+    Function<TaskStatusPlus, Iterable<ResourceAction>> raGenerator = taskStatusPlus -> {
+      final String taskId = taskStatusPlus.getId();
+      final Optional<Task> optionalTask = taskStorageQueryAdapter.getTask(taskId);
```

Comment:
This is still doing `taskStorageQueryAdapter.getTask(taskId)` for every single task. Let's try to structure the code to avoid this somehow. Possibly by retrieving dataSource info upfront for all relevant tasks, in one SQL call, and then caching it.